### PR TITLE
optional storage token argument added to deploy

### PR DESCRIPTION
**Pull request by RuariM**
- [ ] Documentation
- [ ] Tests
- [ ] Code complete
- [ ] Changeset 

This PR is branched from https://github.com/simpleweb/open-format/pull/83.

In this PR I have made it possible to pass a custom NFT storage url. 
A default url will be used if no custom url is passed.
Iv also added a new test to cover this functionality called 'can deploy with an NFTStorage token'

### CODE REVIEW

File: sdks/open-format/src/core/contract.ts [hunk 1]
```diff
@@ -420,6 +415,9 @@ export async function burn({
   return receipt;
 }
 
+const defaultNFTStorageToken =
+  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJkaWQ6ZXRocjoweDVkMjJDZDg2M2Y2NUVBN0ZjZjI3MEE5MUY2NTE5Nzc4OGRhRjU4NmMiLCJpc3MiOiJuZnQtc3RvcmFnZSIsImlhdCI6MTY1OTAxMTU3OTQ2NSwibmFtZSI6Ik9wZW4gRm9ybWF0In0.v-0FkXH0IpVB_JdyG6Ho6kXPqy4DOVr_HZzsks3DES4';
+
```
Comment (benjaminreid) [inline sdks/open-format/src/core/contract.ts:420]: I think we want to remove the default token entirely, so we're not exposing our own token. Basically if they don't provide a token, we don't do the upload and they need to provide their own IPFS URL.

File: sdks/open-format/src/core/contract.ts [hunk 2]
```diff
@@ -454,6 +454,12 @@ export async function deploy({
 
     const customMetadata = nft.metadata ?? {};
 
+    console.log(nftStorageToken);
```
Comment (benjaminreid) [inline sdks/open-format/src/core/contract.ts:457]: ✂️

File: sdks/open-format/src/core/contract.ts [hunk 3]
```diff
@@ -454,6 +454,12 @@ export async function deploy({
...
+
+    const nftStorage = new NFTStorage({
+      token: nftStorageToken || defaultNFTStorageToken,
```
Comment (benjaminreid) [inline sdks/open-format/src/core/contract.ts:460]:
```suggestion
      token: nftStorageToken,
```

After we remove the default token.

File: sdks/open-format/src/core/contract.ts [hunk 4]
```diff
@@ -454,6 +454,12 @@ export async function deploy({
```
Comment (benjaminreid) [inline sdks/open-format/src/core/contract.ts:454]: I think we need another `invariant` line here, something like...

```tsx
invariant(nftStorageToken, 'Please set a NFT storage token, more information https://docs.openformat.simpleweb.co.uk/guides/deploying');
```

This way, it'll stop the `NFTStorage` below failing to be created with an empty token.

**Review by benjaminreid — CHANGES_REQUESTED** [review]
Looks good, just a few things to address 🙏

**Comment by RuariM** [comment]
So I'm assuming I'll need to refractor the others deploy test so they each also pass a storage token? And create a test that makes sure an error is thrown when no token is passed?

**Comment by benjaminreid** [comment]
> Cool. So I'm assuming I'll need to refractor the others deploy test so they each also pass a storage token? And create a test that makes sure an error is thrown when no token is passed?

Yeah, that sounds reasonable to me 👍

One thing that could ease this, and we did mention in the call is being able to set the token when you initialise the SDK (but still make it optional to pass on deploy like you have now.

```tsx
const sdk = new OpenFormatSDK({
  network: 'http://localhost:8545',
  nftStorageToken: "zzz",
  factory: "some-id",
});
```

Then you could call `sdk.deploy` without having to pass the token every time.

**Comment by benjaminreid** [comment]
@RuariM Or actually, some of the deploy tests can have a dummy URL passed to them if it's not important to the test which should skip the NFT Storage upload anyway.

**Comment by RuariM** [comment]
Ah ok that makes sense.

**Comment by RuariM** [comment]
Also what is the purpose of the transactionArgs argument passed to the deploy function?

**Comment by benjaminreid** [comment]
IIRC they are passed to the deploy function on the contact, https://github.com/simpleweb/open-format/blob/main/sdks/open-format/src/core/contract.ts#L445. @tinypell3ts was using it to adjust gas fees or similar I think...

**Comment by RuariM** [comment]
I think the other tests should pass in their current form I will just need to add a test to check that deploy fails without a token and url.

File: sdks/open-format/src/core/contract.ts (hunk 4 above)
Comment (RuariM) [reply sdks/open-format/src/core/contract.ts:454]: > Refe

ah yep I get you.

**Comment by RuariM** [comment]
> IIRC they are passed to the deploy function on the contact, https://github.com/simpleweb/open-format/blob/main/sdks/open-format/src/core/contract.ts#L445. @tinypell3ts was using it to adjust gas fees or similar I think...

Ah, yep I see what you mean.

**Comment by RuariM** [comment]
does it make more sense to pass a file path to the deploy function which then converts it to the File type that nft storage accepts? 

For example inside the deploy function in the contract.ts file.
```
async function fileFromPath(filePath: string) {
    const content = fs.readFileSync(filePath);
    return new File([content], path.basename(filePath), { type: 'image/png' });
}

const metadata = await nftStorage.store({
      name: nft.name,
      description: nft.description,
      image: fileFromPath(nft.image),
      factory_id: factory,
      release_type: nft.releaseType,
      ...customMetadata,
    });
```

Currently a File type is being passed to the function.

**Review by benjaminreid — APPROVED** [review]

**Pull request merged**